src/os/CMakeLists.txt

```diff
@@ -38,7 +38,8 @@ endif(WITH_BLUESTORE)
 if(HAVE_LIBAIO OR HAVE_POSIXAIO)
   list(APPEND libos_srcs
     bluestore/KernelDevice.cc
-    bluestore/aio.cc)
+    bluestore/aio.cc
+    bluestore/io_uring.cc)
```

Comment:
When I look at the `io_uring` code, I do not think it is going to compile work under `HAVE_POSIXAIO`. 
If only because it includes a few Linux specific files.
Not even thinking about doing sycalls() on a not Linux platform.

So accommodating for this would be greatly appriciated. Perhaps that more extreem refactoring will heb untangling this code.

Reply:
Indeed that can be a problem. io_uring supposed to exist along with libaio only. Since libaio is a linux thingy only, so should be fine without introducing another macro.

Reply:
@wjwithagen ping?

Reply:
@tchaikov @rouming 

I think it will currently work, although they way it is constructed is a first and new for me.
The code in the `.cc` is ifdefed out for `NOT HAVE_IO_URING` 

Personally I would prefer that to be implemented in the CMake files.
And it will require some rewrites in #25936 , due to renaming `aio_queue_t`.
So testing this by compilation will take some more time...

Reply:
> Personally I would prefer that to be implemented in the CMake files.

The idea is to avoid spoiling KernerlDevice.cc with ifdefs but have only runtime check where queue is constructed.

Reply:
@rouming
That is going to happen anyways.
Since I also have #25936 in the queue, which is going to have a lot of ifdef's in `KernelDevice.cc` to be able to implement `POSIXAIO`

BTW the current version did compile on FreeBSD/Clang

Reply:
> That is going to happen anyways.

For sure, but at least not because of me :)
Thanks for compiling that on freebsd.

Reply:
@rouming 
That is a nonsense argument...
And not a reason to do it in another way that all other code is done.

Reply:
What argument is nonsense? That I did not use ifdefs in order not to spoil already quite complicated code? I feel it a bit differently: the code which is hard to read and change because of ifdefs has to be refactored or at least not to be inflated with new ones.

Reply:
@rouming 
nonsense:
> For sure, but at least not because of me :)

And yes it is complex code. And the way you did it, will make it even less easy to understand.
Have you looked at #25936 to see what needs to go in as well?
But I'll leave it up to @tchaikov, he'll merge if he sees fit.

Reply:
> Have you looked at #25936 to see what needs to go in as well?

Oh, that's exactly what I tried to avoid: different arguments for similar API calls for different io engines which should be wrapped with ifdef in each caller.  Why not to make a separate posix io engine and make it coexist with libaio or any other engine? Kernel device chooses then proper io engine at runtime according to the configuration

Reply:
@rouming
Because on FreeBSD the other engines really are not available....
And there are no other users of these routines.

And if you really want to make it an API, then I would perhaps expect more abstraction.
Now you called the abstraction `io_{u}ring`.
In a real API I would expect all the `ring` stuff to be hidden down into the `uring` files, and all 3 implementations would be on par. Where as now most code is slightly `uring` biased.
But like I before said: we've now moved into parts where I leave it up to @tchaikov.

Reply:
@wjwithagen thanks for raising this concern. IIRC, @markhpc was not very content about this on how we arrange this and he had concern regarding to the impact of performance.. that's why i want to do some basic perf test on with libaio backend, and i was not quite understand his concern by then. as you're also not satisfied with this, in addition to running the tests, i will revisit this change.

Reply:
@wjwithagen 

> Now you called the abstraction io_{u}ring.

i don't think it's an abstraction. it's more like an implementation of `io_queue_t`.

> In a real API I would expect all the ring stuff to be hidden down into the uring files, 

i cannot find the ring stuff out of `io_uring.{h,cc}`. could you point me to it?

> Where as now most code is slightly `uring` biased.

i am reviewing this PR and #25936. i don't really understand why this change is "slightly `uring` biased`. the io-uring stuff is encapsulated in `io_uring.{h,cc}`. and the aio implementation is not "polluted" by its io uring counterpart. if you want to improve the FreeBSD support of the POSIX AIO backend, i think you can either update `aio_queue_t` or create yet another derived class of `io_queue_t`, if it's difficult to share the code between POSIX AIO and libaio backends.

Reply:
@tchaikov

Maybe I'm just being too pedantic, so like I suggested before, feel free to commit this

Perhaps I just stumbling over the fact that to top level include is `ceph_io_uring.h` and it contains things like:
```
struct ioring_data;

struct ioring_queue_t final : public io_queue_t {
  std::unique_ptr<ioring_data> d;
  unsigned iodepth = 0;

  typedef std::list<aio_t>::iterator aio_iter;

  // Returns true if arch is x86-64 and kernel supports io_uring
  static bool supported();

  ioring_queue_t(unsigned iodepth_);
  ~ioring_queue_t() final;

```
Perhaps is can be as easy as remove all `ring` from the class members and then implement them in the appropriate derived class like you suggested above.

Was the design first skewed toward `aio_` now `ceph_io_uring.h` seems to be the top of the io_queue stuff. That was first by natuaral cause, because there was only on.

And if we want the create three equal implementations than I think it would have an an `ceph_io_queue.h` class with an abstract interface that gets implemented by each of the implementations. Perhaps that the selection between `aiolib` and `io_uring` in the same code might cause a bit more code.

The AIO and POSIX-AIO code arleady as parts where I find it much more legible to `ifdef` large blocks of code, instead of ìfdef`-ing every other line. So that is close to a refactor.

Reply:
i don't understand. ceph_io_uring.h **is** supposed to encapsulate the io uring stuff, right?

> Perhaps is can be as easy as remove all ring from the class members and then implement them in the appropriate derived class like you suggested above.

i am not suggesting it. it is already implemented that way.


> with an abstract interface

that abstract interface is `io_queue_t`, i think.

Reply:
@tchaikov 
Perhaps too much nittpicking, but then I'd expect the include file to be called something like
`io_queue.h` which in itself could include whatever other `io_*.h` files it prefers.